services/api/src/routes/products.js

```diff
@@ -26,13 +28,15 @@ router
     }),
     async (ctx) => {
       const product = await Product.create(ctx.request.body);
+      await Audit.create(ctx, 'create product', product.id, { name: product.name });
       ctx.body = {
         data: product,
       };
     }
   )
   .get('/:productId', async (ctx) => {
     const { product } = await ctx.state;
+    await Audit.read(ctx, 'view product', product.id);
```

Comment:
We're not really reading the audit (report) here nor are we reading the audit trail. We're creating an audit trail entry, so more correct and clear would be:

- `AuditTrail.add('read')`
- `AuditTrail.append('read')`
- `AuditTrailEntry.create('read')`
- `createAuditTrailEntry('read')`
- `appendAuditTrail('read')`

Reply:
I was drawing parallels to logger.error doesn't throw an error but logs an error. Similar i would `Audit[type](something)`

Its not clear to me if you want the just one method for each for the different types or you are not seeing that I got a method for each type (because the interface is different)?

Reply:
I think there's a lot of assumptions baked into the needs here and CRUD style verbs is one. A fair assumption is that we always have an action but that might not fall into CRUD. We have more assumptions too like there will always be a context object and there will always be an ID which I also think may or may not be true.

Given this I think defining our interface less strictly is a good idea so for example:

```
AuditEntry.create('read', {
   ctx, // <-- a "ctx" could be something that AuditEntry knows how to "unpack"
   objectId: product.id,
   userId: ctx.state.authUser.id,
});
```

or even:

```
AuditEntry.create({
   action: 'read',
   objectId: product.id,
   userId: ctx.state.authUser.id,
   ctx,
});
```

Even if `action` is required I think I prefer a single options object to multiple args for readability/composability.

Reply:
yeah i think i am gonna go that way, its gets around the issue disagreement around naming. 

And can be added later.